### ISSUE TITLE
MySql connection error

### ISSUE THREAD

**Issue by mPanasiewicz** [jondot/sneakers#302]
Hello,
I've got some problem on production environment. I start my workers and everything is fine. My sneakers configuration file:
require 'sneakers'
```
Sneakers.configure(amqp: "amqp://#{ENV['RABBITMQ_USER']}:" \
                                 "#{ENV['RABBITMQ_PASS']}@" \
                                 "#{ENV['RABBITMQ_HOST']}:"\
                                 "#{ENV['RABBITMQ_PORT']}",
                   workers: Rails.configuration.rabbitmq_workers,
                   exchange: "promohub_#{ENV['RABBITMQ_ENV']}_messages",
                   threads: 25,
                   prefetch: 25,
                   exchange_type: :topic,
                   heartbeat: 2,
                   env: ENV['RABBITMQ_ENV'])

Sneakers.configure(timeout_job_after: 10) unless Rails.env.development?
Sneakers.logger.level = Logger::INFO # Rails.env.development? ? Logger::DEBUG : Logger::INFO
```
It's working but after few days(about 3-5 days) my worker lost mysql connection.
I set up timeouts on DB(600seconds to cut wrong pools). My db has got 100 pools for 2 workers, 25 threads. It's enough I think.
I'd like to prevent mysql lost connection and set up sth like this:
```
after_fork do
  ActiveSupport.on_load(:active_record) do
    ActiveRecord::Base.establish_connection
  end
  Sneakers.logger.info " !! im after forke'n !! "
end

before_fork do
  ActiveRecord::Base.connection_pool.disconnect!
  Sneakers.logger.info " ** im before-fork'en ** "
end
```

The documentation(https://github.com/jondot/sneakers/wiki/Configuration) said that it is possible but I don't know where should I put my sneakers.conf.rb
The instruction from the documentation is not clear becouse I get undefined method for after_fork or I've got wrong implementation.
I neeed to reconnect mysql connection if sneakers lost this connection somehow. 
Thank you for your help.

**Comment by mPanasiewicz** [comment]
OK my solution is simple. IN my class where I invoke query I set:
```
   def MyClass(message)
        tries ||= 3
        @mpc = Model.find_by(code: message[:x][:y])
      rescue Mysql2::Error => e
        Rails.logger.error(e.message)
        ActiveRecord::Base.clear_active_connections!
        retry unless (tries -= 1).zero?
        raise
      end
```

**Comment by Tensho** [comment]
I've met the same issue in the specific circumstance and I have an assumption regarding the root of the cause. ActiveRecord maps each connection per thread and each sneakers worker (in a separate thread) uses dedicated ActiveRecord connection. That leads us to long-running database connections without check out/check in loop familiar to Rails and Sidekiq. Both gems have special middleware (more accurately, Rails 5+ switched to Executor model) to check out (get) connection from the pool and check in (return) it back after request/job processing. Such approach eliminates the situation when MySQL server drops a connection after timeout of idle and ActiveRecord connection becomes stale.

AFAIK, ActiveRecord either maps each connection pool to process and doesn't require to disconnect from the master process:
```
# config/initializers/sneakers.rb
Sneakers.configure runner_config_file: 'config/sneakers.conf.rb',
                   connection:        Bunny.new,
                   daemonize:         false,
                   workers:           1
```
```
# config/sneakers.conf.rb
before_fork do
  puts "I'm in a parent process! PID##{Process.pid}"
  puts "ActiveRecord::Base.connection_pool.object_id: #{ActiveRecord::Base.connection_pool.object_id}"
end


after_fork do
  puts " I'm in a child process! PID##{Process.pid}"
  puts "ActiveRecord::Base.connection_pool.object_id: #{ActiveRecord::Base.connection_pool.object_id}"
end
```
```
$ bundle exec rake sneakers:run
2018-04-12T13:06:07Z p-21415 t-oxcvsg16o WARN: Loading runner configuration...
2018-04-12T13:06:07Z p-21415 t-oxcvsg16o INFO: Loading config with file: config/sneakers.conf.rb
2018-04-12T13:06:07Z p-21415 t-oxcvsg16o DEBUG: New configuration: #<Sneakers::Configuration:0x007fe9a2ce1e70 ...>
2018-04-12T13:06:07Z p-21415 t-oxcvsg16o WARN: Loading runner configuration...
2018-04-12T13:06:07Z p-21415 t-oxcvsg16o INFO: Loading config with file: config/sneakers.conf.rb
2018-04-12T13:06:07Z p-21415 t-oxcvsg16o DEBUG: New configuration: #<Sneakers::Configuration:0x007fe9aa2042c0 ...>
2018-04-12T13:06:07Z p-21415 t-oxcvsg16o WARN: Loading runner configuration...
2018-04-12T13:06:07Z p-21415 t-oxcvsg16o INFO: Loading config with file: config/sneakers.conf.rb
2018-04-12T13:06:07Z p-21415 t-oxcvsg16o DEBUG: New configuration: #<Sneakers::Configuration:0x007fe9aa23a118 ...>
I'm in a parent process! PID#23301
ActiveRecord::Base.connection_pool.object_id: 70256773274980
 I'm in a child process! PID#23318
ActiveRecord::Base.connection_pool.object_id: 70256802716440
```

@michaelklishin, @jondot If my description makes sense, maybe it's worth to introduce middleware level to sneakers the same way as Sidekiq has? Or at least drop in the check out/check in code around `do_work` method with `ActiveRecord` constant check.

**Comment by michaelklishin** [comment]
I'm sceptical of the idea to introduce anything ActiveRecord-specific to Sneakers. A separate library we'd recommend would be a much better approach.

**Comment by Tensho** [comment]
An external library requires any kind of hook should be introduced to **sneakers** (not middleware mandatory) to keep clean public interfaces over `Sneakers::Worker` patching.

**Comment by mPanasiewicz** [comment]
@michaelklishin how you resolve the problem of mysql lost connections in sneakers workers?? Do you use retries? Or some library??

**Comment by Tensho** [comment]
I keep the next initializer code:
```
module Sneakers
  module ActiveRecord
    module ConnectionManagement
      # https://github.com/jondot/sneakers/issues/302
      # https://github.com/rails/rails/commit/c7b7c6ad1c773102753f1a11b857d0e37ceb6a21
      def process_work(delivery_info, metadata, msg, handler)
        super
      ensure
        ::ActiveRecord::Base.clear_active_connections! unless Rails.env.test?
      end
    end
  end
end

module Sneakers
  module Worker
    logger.warn "Patch Sneakers::Worker to clear ActiveRecord connections"
    prepend Sneakers::ActiveRecord::ConnectionManagement
  end
end
```

**Comment by rg34** [comment]
We inherited a legacy Rails app (Ruby 2.6.4, Rails 5.2.5) and it’s using Sneakers to consume and publish to Rabbitmq. We have 4 distinct jobs and most of them update multiple databases (the Rails app db itself, and 2 others). We have `share_threads: true` and `threads: 10` and this is working great. With kubernetes, we autoscale our “consumers” from 1 - 10 instances.
The concern is where 2 of our databases are on the same server. At peak we’ll have:
4 workers x 10 threads for each of these dbs, so 80 threads total x 10 instances. When pods scale down, those connections are obviously terminated, but on remaining pods, we’ll have 80 threads sleeping for many hours if there are no messages being processed. 

Our netops team isn’t happy that these threads sleep for so long. They have asked that we disconnect these in a reasonable amount of time (under 5 minutes) from the Sneakers side. We do not want to handle this on the Mysql side, using `wait_timeout` for example, since these dbs are accessed from many other applications. But, we also don’t want to limit our threads because we don’t want to sacrifice performance. 

I implemented your initializer @Tensho and it's working great! Thank you. I see that connections are killed after sleeping for 6 minutes and this seems like a viable solution for us. One question:
Can any argument be made from a Sneaker's performance perspective for keeping these threads alive longer or until Mysql kills them (default 8 hours)? 

Thank you so much for your insight.
